### PR TITLE
WIP: Ongoing work to ZIO Test

### PR DESCRIPTION
**Pull request by jdegoes**

### CODE REVIEW

File: test/shared/src/main/scala/zio/test/Predicate.scala [hunk 1]
```diff
@@ -17,6 +17,7 @@
 package zio.test
 
 import zio.Exit
+import scala.reflect.ClassTag
```
Comment (NeQuissimus) [inline test/shared/src/main/scala/zio/test/Predicate.scala]: Unused

File: test/shared/src/main/scala/zio/test/TestAspect.scala [hunk 2]
```diff
@@ -51,6 +55,19 @@ object TestAspect {
       def apply[R >: R0 <: R0, E >: E0 <: E0](test: ZIO[R, E, TestResult]): ZIO[R, E, TestResult] = f(test)
     }
 
+  /**
+   * An aspect that retries a test until success, without limit.
```
Comment (NeQuissimus) [inline test/shared/src/main/scala/zio/test/TestAspect.scala]: I can compose this with a limit?

File: test/shared/src/main/scala/zio/test/package.scala [hunk 3]
```diff
@@ -27,12 +27,13 @@ package zio
  *
  * {{{
```
Comment (NeQuissimus) [inline test/shared/src/main/scala/zio/test/package.scala]: Should we introduce something that type-checks Scaladoc or move these examples into the microsite?

File: test/shared/src/main/scala/zio/test/TestAspect.scala (hunk 2 above)
Comment (jdegoes) [reply test/shared/src/main/scala/zio/test/TestAspect.scala]: Sure, e.g. `retry >>> timeout(60.seconds)`, but if you want something fancier, you're better off using `repeat(schedule)`.

File: test/shared/src/main/scala/zio/test/package.scala (hunk 3 above)
Comment (jdegoes) [reply test/shared/src/main/scala/zio/test/package.scala]: For sure, but let's wait till API stabilizes a bit, probably another week.

File: test/shared/src/main/scala/zio/test/TestAspect.scala (hunk 2 above)
Comment (NeQuissimus) [reply test/shared/src/main/scala/zio/test/TestAspect.scala]: What I figured, just wanted to make sure :)

Comment (jdegoes) [reply test/shared/src/main/scala/zio/test/TestAspect.scala]: You know how I _LOVE_ composability. 😉

File: test/shared/src/main/scala/zio/test/TestAspect.scala [hunk 4]
```diff
@@ -68,6 +68,25 @@ object TestAspect {
...
     }
 
+  /**
+   * An aspect that repeats the test a specified number of times, ensuring it
+   * is stable ("non-flaky").
+   */
+  def nonFlaky(n0: Int): TestAspectPoly =
+    new TestAspectPoly {
+      def apply[R >: Nothing <: Any, E >: Nothing <: Any](test: ZIO[R, E, TestResult]): ZIO[R, E, TestResult] = {
+        def repeat(n: Int): ZIO[R, E, TestResult] =
```
Comment (NeQuissimus) [inline test/shared/src/main/scala/zio/test/TestAspect.scala]: You should be able to put a `@tailrec` here, no?

Comment (jdegoes) [reply test/shared/src/main/scala/zio/test/TestAspect.scala]: It's monadically tail recursive, so the compiler can't accept that as `@tailrec`.

Comment (NeQuissimus) [reply test/shared/src/main/scala/zio/test/TestAspect.scala]: Ah, that's true...

**Pull request merged**